### ISSUE TITLE
Replacing MapCSS map styles by CartoCSS

### ISSUE THREAD

**Issue by Nakaner** [OpenRailwayMap/OpenRailwayMap#663]
As the new maintainer of OpenRailwayMap, I plan to replace the current rendering stack (Kothic.JS, node-tileserver) and map style (the MapCSS one) by a style written in CartoCSS using Mapnik as rendering engine and Tirex and mod_tile as tileserver software. The MapCSS style should serve the needs of JOSM users better while the CartoCSS style can aim to make the best (detailed) railway map based on OSM data.

In the early days of OpenRailwayMap, @rurseekatze started with vector tiles using Kothic.JS as rendering engine and MapCSS as styling language. The main benefit was a smaller tile cache because he could offer multiple map styles with one single set of vector tiles. Unfortunately, vector tiles as they were did not satisfy performance expectations because they were quite big in size. In order get a more performant map (faster loading), he wrote node-tileserver, a NodeJS programme rendering PNG tiles and caching them on server side using Kothic.JS.

More than five years have passed since then. Kothic.JS is unmaintained. Node-Tileserver itself is not able to provide a good label rendering because it cannot render labels beyond the edges of tiles. That's why you see the abbreviations of station names on many zoom levels.

Because our rendering stack is unusual, we cannot reuse existing utilities. For example, printing the map is difficult (you need to dig into Kothic.JS). On the other hand, the raster rendering toolchain based on Mapnik and Tirex is well established (Tirex could be replaced by other software without changing the map style). It has been used for about 10 years now and companies succeeded in making a sustainable business with it.

In order to be able to print OpenRailwayMap, I started reimplementing the infrastructure style in February 2018 in CartoCSS. CartoCSS is a styling language with many similarities to MapCSS at the first glance. It uses almost the same syntax for selectors and properties but CartoCSS comes, like classical GIS, with layers. A YAML file (.mml suffix) describes the data sources (often SQL queries) of theses layers. CartoCSS can be compiled into Mapnik XML styles using Carto (there is a second compile MagnaCarta with limited support but support for Mapserver). I did not invest a lot of time back then but I brought the port to a useable state a few weeks ago and ported the maxspeed and signal styles afterwards.

The ported styles are now ready for a first review (a tile server still needs to be set up). There are a couple of changes because I slightly improved the style a bit:

* All tracks will have a white halo (CartoCSS term: casing) to be easier to distinguish from background maps.
* Narrow gauge railways do not have dashes any more because they are difficult to recognize and look poor if there is more than one track in parallel. Instead, all narrow gauge tracks got a new yellowish colour. `usage=main` has no effect on these tracks.
* Bridge rendering is copied from OSM Carto.
* Symbols for radio towers and antennas got a higher priority. Labels of milestones and line names have a lower priority now but still appear often enough.
* Milestones with non-integer distances will be rendered on higher zoom levels only.
* Operating sites with passenger traffic are sorted by their importance using a couple of SQL statements. The importance is calculated based on the number of route relations their platforms and stop positions belong to. Relations of `type=public_transport` + `public_transport=stop_area` are not taken into account because they are not available in a database imported with Osm2pgsql. The importance is pre-calculated and stored in a materialized view because it requires access to the `planet_osm_rels` table and is not very performant due to usage of array fields in that table.
* Signal boxes mapped as points and – on smaller zoom levels only – polygons are rendered as green dots with a white halo.
* All styles use Noto Sans (fallback: Unifont) as font. Noto Sans has a very wide range of supported scripts and is used by OSM Carto and many other map styles as well.
* All symbols are SVGs.
* Rendering of speed limits of tracks with different limits in both directions is reimplemented. Labels indicate the dominant direction of traffic based on `railway:preferred_direction=*`. This direction influences the colouring as well. Labels are either "140/100" for a track with 140 in one and 100 kph in the other direction and both as preferred directions, or "140 (100)" if the direction with 140 kph speed limit is the preferred direction. Missing speeds are indicated as dashes, e.g. "140 (–)".
* Razed, abandoned, disused, proposed railways and railways under construction are not rendered on low zoom levels if their lifecycle tags indicating their future/old type (e.g. `disused:railway=tram`) are missing. This avoids cluttering with many abandoned tram lines in cities and allows to render abandoned railway lines on lower zoom levels than now.
* LZB Blockkennzeichen are rendered with their reference number (like road shields in road map styles)

The source code is located in a separate repository because it is not directly related with the frontend.

You can review a couple of static images at
https://michreichert.de/share/public/orm-cartocss-examples.zip (33 MB)
The ZIP archive contains all three styles (but some locations with signal style only if they have special signals to demonstrate). Please view the images at 100% in your image viewer.

The centre locations of the images are

```
9.177 48.7803 (Stuttgart Hbf (tief))
8.4022 48.9936 (Karlsruhe Hbf)
8.2704 49.0468 (Wörth (Rhein))
13.2874 51.3096 (Riesa)
10.0075 53.5517 (Hamburg Hbf)
9.9259 51.5366 (Göttingen)
6.6859 51.2118 (Neuss Gbf)
11.1237 50.9946 (Vieselbach Umspannwerk)
```
